lib/execution.go

```diff
@@ -0,0 +1,562 @@
+// of gracefullt winding down.
+//
+// IMPORTANT: for UI/information purposes only, don't use for synchronization.
+func (es *ExecutionState) GetCurrentlyActiveVUsCount() int64 {
+	return atomic.LoadInt64(es.activeVUs)
+}
+
+// ModCurrentlyActiveVUsCount changes the total number of currently active VUs.
+//
+// IMPORTANT: for UI/information purposes only, don't use for synchronization.
+func (es *ExecutionState) ModCurrentlyActiveVUsCount(mod int64) int64 {
+	return atomic.AddInt64(es.activeVUs, mod)
+}
+
+// GetFullIterationCount returns the total of full (i.e uninterrupted) iterations
+// that have been completed so far.
+//
+// IMPORTANT: for UI/information purposes only, don't use for synchronization.
+func (es *ExecutionState) GetFullIterationCount() uint64 {
+	return atomic.LoadUint64(es.fullIterationsCount)
+}
+
+// AddFullIterations increments the number of full (i.e uninterrupted) iterations
+// by the provided amount.
+//
+// IMPORTANT: for UI/information purposes only, don't use for synchronization.
+func (es *ExecutionState) AddFullIterations(count uint64) uint64 {
+	return atomic.AddUint64(es.fullIterationsCount, count)
+}
+
+// GetPartialIterationCount returns the total of partial (i.e interrupted)
+// iterations that have been completed so far.
+//
+// IMPORTANT: for UI/information purposes only, don't use for synchronization.
+func (es *ExecutionState) GetPartialIterationCount() uint64 {
+	return atomic.LoadUint64(es.interruptedIterationsCount)
+}
+
+// AddInterruptedIterations increments the number of partial (i.e interrupted)
+// iterations by the provided amount.
+//
+// IMPORTANT: for UI/information purposes only, don't use for synchronization.
+func (es *ExecutionState) AddInterruptedIterations(count uint64) uint64 {
+	return atomic.AddUint64(es.interruptedIterationsCount, count)
+}
+
+// MarkStarted saves the current timestamp as the test start time.
+//
+// CAUTION: Calling MarkStarted() a second time for the same execution state will
+// result in a panic!
+func (es *ExecutionState) MarkStarted() {
+	if !atomic.CompareAndSwapInt64(es.startTime, 0, time.Now().UnixNano()) {
+		panic("the execution scheduler was started a second time")
+	}
+}
+
+// MarkEnded saves the current timestamp as the test end time.
+//
+// CAUTION: Calling MarkEnded() a second time for the same execution state will
+// result in a panic!
+func (es *ExecutionState) MarkEnded() {
+	if !atomic.CompareAndSwapInt64(es.endTime, 0, time.Now().UnixNano()) {
+		panic("the execution scheduler was stopped a second time")
+	}
+}
+
+// HasStarted returns true if the test has actually started executing.
+// It will return false while a test is in the init phase, or if it has
+// been initially paused. But if will return true if a test is paused
+// midway through its execution (see above for details regarind the
+// feasibility of that pausing for normal executors).
+func (es *ExecutionState) HasStarted() bool {
```

Comment:
I see you changed this from `IsRunning()`.

I slightly prefer using the present tense in these cases (`IsRunning()`, `IsFinished()`, etc.), because you want to know the _current_ state, not if something happened in the past that remains true now. And this would align with `IsPaused()`.

Anyways, we could bikeshed about this, but disregard if you prefer the current convention. :smile:

Reply:
```  But if will return true if a test is paused
// midway through its execution (see above for details regarind the
// feasibility of that pausing for normal executors).```
So while a couple of  typos (including an `if->it`)  it has the meaning of it has started at all, not if it is currently running :)

Reply:
Hhmm right the meaning is different, and the current implementation has no need for `IsRunning()`.

OK, disregard. :)